Allow testing option tags with test_creation

nilsdeppe opened the pull request:
## Proposed changes

- Move `test_*_creation` into the `TesHelpers` namespace
- Extend `test_*_creation` to allow testing  the option tags.
- Remove the two extra spaces at the beginning of the option strings passed to the `test_*_creation` functions.

I'm not sure what we are currently supposed to do for testing option tags, so I decided to see what people think about this. One question I have: @wthrowe what do you think is the best way to test this code? Maybe adding a `Test_TestCreation.cpp`?

Commit order:
1. Move test_*_creation into TestHelpers namespace
2. Add tests for test_creation and option tag as template param
3. Remove extra spaces in test_*_creation calls

### Types of changes:

- [ ] Bugfix
- [x] New feature

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

fmahebert commented:
I think I'm missing some context — when you say "testing option tags", what do you mean? Can you give an example of what such a test would look like?

nilsdeppe commented:
What I've used it for is:

```cpp
const ReconstructionDispatch dispatcher_from_opts =
    test_creation<ReconstructionDispatch, void,
                  OptionTags::ReconstructionScheme>(input_file_string);
```

Basically verifying that the option tag is compliant with the Options interfaces and gives the desired result.

wthrowe commented:
This seems like a reasonable direction to go in, and I think a Test_TestCreation.cpp makes sense.  I'm not going to do a detailed review because, according to Travis, the current version doesn't work.

nilsdeppe commented:
Thanks @wthrowe! That was all I was looking for initially. I'll add the test and clean up the rest of the code. I'll move the `test_*creation` functions into the `TestHelpers` namespace as well. I'll ping once that's ready :)

nilsdeppe committed:
Move test_*_creation into TestHelpers namespace

nilsdeppe committed:
Add tests for test_creation and option tag as template param

- Add a bunch of tests for test_*_creation with groups, etc.
- Add the ability to specify an option tag that can be used to parse the string
  to the test_*_creation functions.

nilsdeppe committed:
Remove extra spaces in test_*_creation calls

nilsdeppe commented:
@fmahebert @wthrowe I've updated the code and the PR description comment to reflect the changes now. Something like 50% of this is white space changes, with another 20% being moving the functions into the `TestHelpers` namespace, then there are ~400 lines of tests for the functions. I think I covered all of the cases in which these functions would be used in the test, but please let me know if you can think of any I missed!

wthrowe commented on tests/Unit/TestCreation.hpp:
```diff
@@ -3,50 +3,103 @@
+struct AddGroups {
+  static std::string apply(const std::string& construction_string) noexcept {
+    return prefix_group<Tag>(construction_string);
+  }
```
What is this function for?  It just forwards to the other function.

nilsdeppe replied:
I've "inlined" the `prefix_group` function here adding `OptionTag` template parameter to `apply`. The reason for having the `prefix_group` function was to not duplicate code (all the string manipulation), but the default template parameter is a much cleaner implementation I think. Thanks!

wthrowe commented on tests/Unit/TestCreation.hpp:
```diff
@@ -3,50 +3,103 @@
+/// Option tags can be tested by passing them as the second template parameter.
+/// If the metavariables are required to create the class then the metavariables
+/// must be passed as the third template parameter. The default option tag is
+/// `TestCreationOpt<T>`.
```
`TestCreationOpt<std::unique_ptr<BaseClass>>`

nilsdeppe replied:
Should be fixed, thanks for catching this!

wthrowe replied:
Try again.

nilsdeppe replied:
done...

fmahebert commented on tests/Unit/TestCreation.hpp:
```diff
@@ -3,50 +3,99 @@
-/// must be indented.
-template <typename T, typename Metavariables = NoSuchType>
+/// Construct an object from a given string. The string must not contain the tag
+/// name.
```
Can you clarify what "the tag" is? Do you perhaps mean "the class name"? E.g. in the test, you create a `ClassWithoutMetavariables` but there is no "Tag" (in the Databox sense) for this class.

Perhaps making a Doxygen snippet from the tests would be useful here. Alternatively, you could note in the documentation that the test gives many examples.

This comment also applies to `test_factory_creation` and `test_enum_creation`.

fmahebert commented on tests/Unit/Test_TestCreation.cpp:
```diff
@@ -0,0 +1,411 @@
+            .value == 6);
+  CHECK(
+      TestHelpers::test_creation<ClassWithoutMetavariables>("SizeT: 7").value ==
+      7);
```
nitpick: suggest moving this simpler case above the earlier three cases, because this could make it a bit easier to read for someone looking here to supplement the documentation.

wthrowe commented:
I am happy with the responses to my comments.

kidder commented:
@nilsdeppe what needs to be done for this PR besides a rebase and squash?

nilsdeppe commented:
Okay, finally updated this... I addressed @fmahebert's comments and rebased on develop

fmahebert commented on tests/Unit/TestCreation.hpp:
```diff
@@ -46,8 +46,13 @@ struct TestCreationOpt {
+/// The string must not contain the
+/// name of the option (which is the name of the struct by default if no
+/// `name()` function is present). For example, to create a struct with a
+/// `SizeT` named option the following would be used:
```
I remain confused... you wrote "the string must not contain the name of the option", but the example has an option `SizeT` and you've used the string "SizeT: 7". As I understand this, you've put the name of the option in the string. 

Is my understanding correct that in this snippet you are constructing an object of type `ClassWithoutMetavariables` where you are giving the option `SizeT` the value 7? Is it specifically "ClassWithoutMetavariables" that should be omitted from the string? If so, I would not call it an option, because as I understand our code `ClassWithoutMetavariables` is some type creatable from options, but is not itself an option.

nilsdeppe replied:
Okay, clearly this needed work... I pushed a fixup that hopefully answers these questions :)

fmahebert commented on tests/Unit/TestCreation.hpp:
```diff
@@ -63,8 +68,13 @@ T test_creation(const std::string& construction_string) noexcept {
+/// name of the option (which is the name of the struct by default if no
+/// `name()` function is present). For example, to create a struct with a
+/// `SizeT` named option the following would be used:
+/// \snippet Test_TestCreation.cpp size_t_argument
```
This is factory creation; why not demonstrate the parsing with a derived class? e.g. line 292ish of the test.

fmahebert commented on tests/Unit/TestCreation.hpp:
```diff
@@ -80,8 +90,13 @@ std::unique_ptr<BaseClass> test_factory_creation(
+/// name of the option (which is the name of the struct by default if no
+/// `name()` function is present). For example, to create a struct with a
+/// `SizeT` named option the following would be used:
+/// \snippet Test_TestCreation.cpp size_t_argument
```
This is enum creation; why not demonstrate the parsing with an enum? e.g. line 387ish of the test.

fmahebert reviewed (commented):
I am phrasing everything as questions because I don't really understand the option parsing... so I have no certainty about what the docs should say, but lots of questions about things I don't understand 😛.

fmahebert commented on tests/Unit/TestCreation.hpp:
```diff
@@ -48,10 +48,12 @@ struct TestCreationOpt {
-/// name of the option (which is the name of the struct by default if no
-/// `name()` function is present). For example, to create a struct with a
-/// `SizeT` named option the following would be used:
+/// The string must not contain the name of the option (specifically, the class
```
Should "class" -> "struct" for consistency with use of "struct" later in this sentence?

fmahebert commented on tests/Unit/TestCreation.hpp:
```diff
@@ -48,10 +48,12 @@ struct TestCreationOpt {
+/// The string must not contain the name of the option (specifically, the class
+/// being created from options, which is the name of the struct by default if
+/// no `name()` function is present). For example, to create a struct named
+/// `ClassWithoutMetavariables` with a `SizeT` named option the following would
```
Should "SizeT named option" -> "option tag SizeT" or similar? (I think in the call earlier we agreed on wording at least)

fmahebert commented on tests/Unit/TestCreation.hpp:
```diff
@@ -70,11 +72,15 @@ T test_creation(const std::string& construction_string) noexcept {
-/// `name()` function is present). For example, to create a struct with a
-/// `SizeT` named option the following would be used:
-/// \snippet Test_TestCreation.cpp size_t_argument
+/// The string must contain the name of the option (specifically, the class
```
again should class -> struct ?

fmahebert commented on tests/Unit/TestCreation.hpp:
```diff
@@ -70,11 +72,15 @@ T test_creation(const std::string& construction_string) noexcept {
+/// being created from options, which is the name of the struct by default if
+/// no `name()` function is present). Note that it is not the name of the base
+/// class, but the name of the derived class that must be given. For example, to
+/// create a pointer of `BaseClass` pointing to a derived class of type
```
Here I suggest "a pointer to `BaseClass`" or "a pointer of type `BaseClass`" or maybe just "a `BaseClass*`" ... but feel free to disagree

fmahebert commented on tests/Unit/TestCreation.hpp:
```diff
@@ -70,11 +72,15 @@ T test_creation(const std::string& construction_string) noexcept {
+/// no `name()` function is present). Note that it is not the name of the base
+/// class, but the name of the derived class that must be given. For example, to
+/// create a pointer of `BaseClass` pointing to a derived class of type
+/// `size_t_argument_base` with a `SizeT` named option the following would be
```
again, what about "option tag" ?

fmahebert reviewed (commented):
Between the improved wording, and our discussion in the meeting earlier today, I find this much clearer now. Here are a few final nitpicks to wording. Please squash the previous fixups and any changes you make in response to these nitpicks.

Thanks for putting in the time to improve the clarity here, I think this will be useful to myself and hopefully others too down the line :)

nilsdeppe commented:
Okay, pushed new fixup and rebased on develop :)

fmahebert commented:
I am happy for you to squash 👍

nilsdeppe commented:
Squashed and rebased :)

kidder commented:
@wthrowe please look at

wthrowe commented on tests/Unit/TestCreation.hpp:
```diff
@@ -12,43 +13,110 @@
+///
+/// The string must contain the name of the option (specifically, the struct
+/// being created from options, which is the name of the struct by default if
+/// no `name()` function is present). Note that it is not the name of the base
```
`name()` overrides don't work with factories.  Implementing it has been discussed in #1611, but nothing has happened yet.

wthrowe commented on tests/Unit/TestCreation.hpp:
```diff
@@ -12,43 +13,110 @@
+template <typename T, typename OptionTag = TestCreationOpt<T>,
+          typename Metavariables = NoSuchType,
           Requires<std::is_enum<T>::value> = nullptr>
 T test_enum_creation(const std::string& enum_string) noexcept {
```
This function is now identical to `test_creation`.  (And its documentation claims that it is not.)

nilsdeppe replied:
I'm really not sure what you want me to change. Could you please clarify?

wthrowe replied:
Probably just remove the function, as it is identical to the more common testing function above (except restricted to only work on enums).  If you would prefer not to do that, change the documentation to correctly describe the implemented behavior.

nilsdeppe replied:
Thanks for the clarification!

nilsdeppe commented:
@wthrowe I've pushed a fixup commit. Thanks for catching the redundancy is the enum testing function!

wthrowe commented:
Looks good.  Squash, etc.

nilsdeppe committed:
Fix clang-tidy false positive in Parallel/Serialize.hpp

nilsdeppe commented:
rebased on develop and squashed.

I added a commit that silences a clang-tidy false positive in `Parallel/Serialize.hpp` and adds `noexcept` to the functions there.

nilsdeppe commented:
Travis failures are the usual timeouts.
